python/lsst/rubintv/production/heartbeat.py

```diff
@@ -0,0 +1,76 @@
+# This file is part of rubintv_production.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import time
+
+from .rubinTv import Uploader
+
+__all__ = ['Heartbeater']
+
+
+class Heartbeater():
+    """A class for uploading heartbeats to the GCS bucket.
+
+    Call ``beater.beat()`` as often as you like. Files will only be uploaded
+    once ``self.uploadPeriod`` has elapsed, or if ``ensure=True`` when calling
+    beat.
+
+    Parameters
+    ----------
+    handle : `str`
+        The name of the channel to which the heartbeat corresponds.
+    uploadPeriod : `float`
+        The time, in seconds, which must have passed since the last successful
+        heartbeat for a new upload to be undertaken.
+    flatlinePeriod : `float`
+        If a new heartbeat is not received before the flatlinePeriod elapses,
+        in seconds, the channel will be considered down.
+    """
+    def __init__(self, handle, uploadPeriod, flatlinePeriod):
+        self.handle = handle
+        self.uploadPeriod = uploadPeriod
+        self.flatlinePeriod = flatlinePeriod
+
+        self.lastUpload = -1
+        self.uploader = Uploader()
+
+    def beat(self, ensure=False, customFlatlinePeriod=None):
+        """Upload the heartbeat if enough time has passed or ensure=True.
+
+        customFlatlinePeriod implies ensure, as long forecasts should always be
+        uploaded.
+
+        Parameters
+        ----------
+        ensure : `str`
+            Ensure that the heartbeat is uploaded, even if one has been sent
+            within the last ``uploadPeriod``.
+        customFlatlinePeriod : `float`
```

Comment:
I'm not sure I understand this parameter.  Above, you use it in the catchup script, but it's `flatlinePeriod`, not `customFlatlinePeriod`.  I'm also not seeing where the flatline period is being used.  Is there a function I'm missing that checks `lastUpload` against `flatlinePeriod` (either the class or custom one), and signals that there's an error?

Reply:
Ah, good catch - that's a typo, that should indeed be `customFlatlinePeriod` in those calls, thanks. Locally it's being used to set the forecast, _i.e._ the time after which the channel is considered down.

On the website, the heartbeats' timestamps are being checked that they're always coming before this "considered-dead time", so that you can keep sending them and pushing that time out, but can also have long-running things, where you say "You won't hear from me for an hour, but that's OK in this instance because I'm doing something very long & slow".

Reply:
Ok, the polling code is elsewhere, that's why I didn't see it.  Makes sense.  I have no other issues, so will mark this as approved.